### PR TITLE
Expand mixer background material.

### PR DESCRIPTION
**Pull request by geeknoid**
This is the first modern write up that describes the mixer we're actually building...

You can see a staged version of this doc here: https://geeknoid.github.io/istio.github.io/docs/concepts/mixer.html

### CODE REVIEW

File: docs/concepts/mixer.md [hunk 1]
```diff
@@ -7,40 +7,210 @@ title: Mixer
...
 {% capture overview %}
-The page explains the role and composition of the Istio mixer.
+The page explains the Istio mixer's role and general architecture.
 {% endcapture %}
 
 {% capture body %}
-## General usage
+## Background
 
 The mixer provides the control-plane abstractions necessary for most real-world multi-tenant services.
```
Comment (douglas-reid) [inline docs/concepts/mixer.md]: Can we assume the audience knows what control-plane abstractions and multi-tenant services are?  Should we provide links to wiki or similar here?

File: docs/concepts/mixer.md [hunk 2]
```diff
@@ -7,40 +7,210 @@ title: Mixer
...
 proceeds to repackage and redirect the data towards configured backends.
 
-Services within the Istio mesh can also directly integrate with the mixer. For example, services may wish to provide rich telemetry for particular operations
-beyond what the proxy automatically collects. Or services may use the mixer for resource-oriented quota management. Services that leverage the mixer in this 
-way are abstracted from environment-specific control plane details, greatly easing the process of hosting the code in
-different environments (different clouds & on-prem)
+Services within the Istio mesh can also directly integrate with the mixer. For example, services may wish to provide rich telemetry
+for particular operations beyond what the proxy automatically collects. Or services may use the mixer for resource-oriented quota
+management. Services that leverage the mixer in this way are abstracted from environment-specific control plane details, greatly
+easing the process of  hosting the code in different environments (different clouds & on-prem)
```
Comment (douglas-reid) [inline docs/concepts/mixer.md]: nit: extra space between in "of  hosting"

File: docs/concepts/mixer.md [hunk 3]
```diff
@@ -7,40 +7,210 @@ title: Mixer
...
+<img style="display:block;width:60%;margin:auto;" src="./mixer/ingress.svg" alt="Flow of inbound traffic to a service." />
+<p style="text-align:center;">Ingress Traffic</p><br/>
+
+<img style="display:block;width:60%;margin:auto;" src="./mixer/egress.svg" alt="Flow of outbound traffic to a service." /> 
+<p style="text-align:center;">Egress Traffic</p>
 
 The mixer provides three core features:
 
-**Precondition Checking**. Enables callers to verify a number of preconditions before responding to an incoming request from a service consumer. 
+- **Precondition Checking**. Enables callers to verify a number of preconditions before responding to an incoming request from a service consumer.
```
Comment (douglas-reid) [inline docs/concepts/mixer.md]: In the preceding text, we refer to this as "policy decisions". We probably should be consistent.

File: docs/concepts/mixer.md [hunk 4]
```diff
@@ -7,40 +7,210 @@ title: Mixer
...
+<p style="text-align:center;">Egress Traffic</p>
 
 The mixer provides three core features:
 
-**Precondition Checking**. Enables callers to verify a number of preconditions before responding to an incoming request from a service consumer. 
+- **Precondition Checking**. Enables callers to verify a number of preconditions before responding to an incoming request from a service consumer. 
 Preconditions can include whether the service consumer is properly authenticated, is on the service's whitelist, passes ACL checks, and more.
 
-**Telemetry Reporting**. Enables services to produce logging, monitoring, tracing and billing streams intended for the service producer itself as well as 
+- **Telemetry Reporting**. Enables services to produce logging, monitoring, tracing and billing streams intended for the service producer itself as well as
```
Comment (douglas-reid) [inline docs/concepts/mixer.md]: should we drop tracing and billing from the text now, and add it back in later?

File: docs/concepts/mixer.md [hunk 5]
```diff
@@ -7,40 +7,210 @@ title: Mixer
...
 
 The mixer provides three core features:
 
-**Precondition Checking**. Enables callers to verify a number of preconditions before responding to an incoming request from a service consumer. 
+- **Precondition Checking**. Enables callers to verify a number of preconditions before responding to an incoming request from a service consumer. 
 Preconditions can include whether the service consumer is properly authenticated, is on the service's whitelist, passes ACL checks, and more.
 
-**Telemetry Reporting**. Enables services to produce logging, monitoring, tracing and billing streams intended for the service producer itself as well as 
+- **Telemetry Reporting**. Enables services to produce logging, monitoring, tracing and billing streams intended for the service producer itself as well as 
 for its consumers.
```
Comment (douglas-reid) [inline docs/concepts/mixer.md]: I would add that they can be used by cluster operators to handle production responsibilities, etc.

File: docs/concepts/mixer.md [hunk 6]
```diff
@@ -7,40 +7,210 @@ title: Mixer
...
 tool to provide some fairness between service consumers when contending for limited resources.
 
+## Attributes
+
+Istio uses *attributes* to control the runtime behavior of services running in the mesh. Attributes are named and typed pieces of metadata
+describing ingress and egress traffic and the environment this traffic occurs in. An Istio attribute carries a specific piece
+of information such as the error code of an API request, the latency of an API request, or the
+original IP address of a TCP connection. Here are a few examples of attributes:
+
+	request.path: xyz/abc
```
Comment (douglas-reid) [inline docs/concepts/mixer.md]: should we use markdown blockquotes or code blocks here?

File: docs/concepts/mixer.md [hunk 7]
```diff
@@ -7,40 +7,210 @@ title: Mixer
...
 
-**Telemetry Reporting**. Enables services to produce logging, monitoring, tracing and billing streams intended for the service producer itself as well as 
+- **Telemetry Reporting**. Enables services to produce logging, monitoring, tracing and billing streams intended for the service producer itself as well as 
 for its consumers.
 
-**Quota Management**. Enables services to allocate and free quota on a number of dimensions, Quotas are used as a relatively simple resource management 
+- **Quota Management**. Enables services to allocate and free quota on a number of dimensions, Quotas are used as a relatively simple resource management 
 tool to provide some fairness between service consumers when contending for limited resources.
 
+## Attributes
```
Comment (douglas-reid) [inline docs/concepts/mixer.md]: i feel like this should link to a more detailed Istio-level discussion of attributes. Attributes are used by mixer for sure, but the proxy and/or mixerclient bits should also talk about them.

Also, I think we should link to existing docs on well-known attributes somewhere.

File: docs/concepts/mixer.md [hunk 8]
```diff
@@ -7,40 +7,210 @@ title: Mixer
...
+ 
+- **Resolution**. The second phase is to evaluate the set of attributes to determine the effective 
+configuration to apply for the request. See *TBD* for information on how resolution works. The effective
+configuration determines the set of aspects and descriptors available to handle the request in the
+subsequent phases.
+
+- **Attribute Processing**. The third phase takes the set of incoming request attributes
+and produces a set of policy objects. Attribute processing is initially
+configured through a simple declarative form as described in *TBD*. Advanced attribute processing can also
+be implemented through a general-purpose scripting environment as described [below](#scripting).
```
Comment (douglas-reid) [inline docs/concepts/mixer.md]: I don't think we should mention "advanced attribute processing" yet. We don't have a complete story for it and I don't think we want to advertise it before we are ready.

File: docs/concepts/mixer.md [hunk 9]
```diff
@@ -7,40 +7,210 @@ title: Mixer
...
+
+- **Adapter Dispatching**. The Resolution phase establishes the set of available aspects and the Attribute
+Processing phase creates a set of policy objects. The Adapter Dispatching phase invokes the adapters
+associated with each aspect and passes them policy objects.
```
Comment (douglas-reid) [inline docs/concepts/mixer.md]: nit: they aren't always `policy objects`. I might rephrase this as "configured parameters".

File: docs/concepts/mixer.md [hunk 10]
```diff
@@ -7,40 +7,210 @@ title: Mixer
...
+set of adapters (mixer plugins) that are responsible for introducing new attributes. These attributes are combined with the attributes
+from the request to form the total set of attributes for the operation.
+ 
+- **Resolution**. The second phase is to evaluate the set of attributes to determine the effective 
+configuration to apply for the request. See *TBD* for information on how resolution works. The effective
+configuration determines the set of aspects and descriptors available to handle the request in the
+subsequent phases.
+
+- **Attribute Processing**. The third phase takes the set of incoming request attributes
+and produces a set of policy objects. Attribute processing is initially
```
Comment (douglas-reid) [inline docs/concepts/mixer.md]: are logs or metrics considered `policy objects`? isn't this really "configured inputs" or "data structures"?

File: docs/concepts/mixer.md [hunk 11]
```diff
@@ -7,40 +7,210 @@ title: Mixer
...
+be implemented through a general-purpose scripting environment as described [below](#scripting).
+
+- **Adapter Dispatching**. The Resolution phase establishes the set of available aspects and the Attribute
+Processing phase creates a set of policy objects. The Adapter Dispatching phase invokes the adapters
+associated with each aspect and passes them policy objects.
+
+<img style="display:block;width:40%;margin:auto;" src="./mixer/phases.svg" alt="Phases of mixer request processing" />
+<p style="text-align:center;">Request Phases</p>
+
+## Configuration state
```
Comment (douglas-reid) [inline docs/concepts/mixer.md]: This section and the "Descriptors and policy objects" should come before Request Phases. They are similar to Attributes in that they describe what is being manipulated. After that, transition to the how seems more natural.

File: docs/concepts/mixer.md [hunk 12]
```diff
@@ -7,40 +7,210 @@ title: Mixer
...
+that conform to the descriptor's shape (it's value type and its set of labels).
+
+- It enables type checking of the deployment's configuration. Since attributes have strong types, and so do descriptors,
+Istio can provide a number of strong correctness guarantees of the system's configuration. Basically, if a chunk of
+configuration is accepted into the Istio system, it means the configuration passes a minimum correctness bar. Again, this
+plays the same role as types in a programming language.
+
+- It enables us to provide a strongly-typed scripting environment.
+
+## Scripting
```
Comment (douglas-reid) [inline docs/concepts/mixer.md]: Again, I really don't think we want this yet. We don't support it all currently, and haven't figured out exactly how it is going to work, etc.

File: docs/concepts/mixer.md [hunk 13]
```diff
@@ -7,40 +7,210 @@ title: Mixer
...
+control to individual configured aspects.
+
+For common uses, the operator authors policy objects production rules via a relatively simple declarative format
+and expression syntax. The mixer ingests such rules and produces a script that performs the necessary runtime work
+of accessing the request's incoming attributes and producing the requisite policy objects.
+
+For advanced uses, the operator can bypass the declarative format and author directly in the scripting
+language. This is more complex, but provides ultimate flexibility.
+
 ## Adapters
```
Comment (douglas-reid) [inline docs/concepts/mixer.md]: I would also move this section above the "Request phases" bit.  Get all the concepts out of the way before talking about how they are woven together.

**Comment by rshriram** [comment]
I will leave the content section to the mixer folks. With regard to the first two diagrams, I suggest merging the two, and numbering the arrows (1-..). You could then refer to the arrows in your discussion. 

Secondly, you could create a separate picture that shows service to mixer communication instead of having both service->mixer and proxy->mixer in one picture. It delivers the impression that the normal control flow is super mixer heavy.

File: docs/concepts/mixer.md (hunk 1 above)
Comment (geeknoid) [inline docs/concepts/mixer.md]: Added text to clarify.

**Comment by geeknoid** [comment]
PTAL.

File: docs/concepts/mixer.md [hunk 14]
```diff
@@ -1,46 +1,200 @@
...
+- a 64-bit integer metric value 
+- a source string
+- a target string
+- a service string
+- a method string
+- a 64-bit integer response code
+
+Here is an example snippet of Istio configuratin which produces a policy object for the above descriptor:
+
+      metrics:
```
Comment (ZackButcher) [inline docs/concepts/mixer.md]: Same, this is a single element list, not a single `MetricAspectConfigParams_Metric` (or whatever the type of that message inside of the aspect params is).

**Comment by geeknoid** [comment]
All issues addressed. PTAL.

Easier reading: 

https://geeknoid.github.io/istio.github.io/docs/concepts/mixer.html
https://geeknoid.github.io/istio.github.io/docs/concepts/attributes.html
https://geeknoid.github.io/istio.github.io/docs/reference/attribute-vocabulary.html
